### PR TITLE
Adding the script to download and process the GPCC file

### PR DESCRIPTION
**Pull request by CamSalda**
This GPCC file is monstrous: it has more than 40 million rows. The process to match latitudes and longitudes is very intensive in terms of computing power and my laptop crashed like 4 times trying to do it.

Finally I decided just to try the code with a smaller sample and it works, so it would be amazing if someone else could try it to be sure if it works with the full file.

Closes https://github.com/jessegrabowski/laos-climate-change/issues/20

### CODE REVIEW

**Comment by jessegrabowski** [comment]
We don't have a super high level of geographic resolution on our other datasets, notably EM-DAT, so we're going to have to summarize down the data. My suggestion would be to loop country by country and compute its average value.

**Comment by CamSalda** [comment]
I agree, but to have the country values, we have to match the latitude and longitude data with the shapely.Points and then geo merge the shapefile with the GPCC monster file, and there is where my computer explodes.

Tomorrow my new computer arrives, so maybe I will be able to do it.

**Comment by CamSalda** [comment]
> We don't have a super high level of geographic resolution on our other datasets, notably EM-DAT, so we're going to have to summarize down the data. My suggestion would be to loop country by country and compute its average value.

The draft of the function is ready for review.

File: laos_gggi/GPCC_data_loader.py [hunk 1]
```diff
@@ -2,81 +2,111 @@
...
+                "ISO_A3": "country_code",
+                "FORMAL_EN": "country",
+                "CONTINENT": "continent",
+                "REGION_UN": "region",
+            }
+        )
+        # Open gpcc files, transform them to geopandas shapefile geometry and merge them with the world shapefiles
         for x in list(path_to_GPCC_raw.keys()):
-            exec(f'data_{x} = xr.open_dataset(output_path + "/gpcc/gpcc_raw_" + x + ".nc")')
+            exec(
```
Comment (jessegrabowski) [inline laos_gggi/GPCC_data_loader.py:82]: Using `exec` should be avoided in general, and it doesn't seem necessary here. You can build up your results incrementally instead:

```
result_df = None
for key in path_to_GPCC_raw:
    data = xr.open_dataset(output_path + "/gpcc/gpcc_raw_" + x + ".nc")'
    df = data_{x}["precip"].to_dataframe().reset_index()
    df_gep = geo.GeoDataFrame(df_{x}, geometry=geo.points_from_xy(df_{x}["lat"], df_{x}["lon"]), crs="EPSG:4326")'

   # ...

   result_df = df if result_df is None else pd.concat([result_df, df], axis=0)

(result_df
    .reset_index()
    .pivot_table(values="precip", index=["country_code", "time"], aggfunc="mean")
    .to_csv(gpcc_processed_path)
)
```

**Review by jessegrabowski — CHANGES_REQUESTED** [review]
Looks good. In general avoid `exec`, in 99% of cases there's a way around it. Also if you find yourself with a lot of copy/pasted code with just one little thing changed, you should definitely turn it into a function. Even if you only call it twice, it reduces the change for an error by 50% and makes the code easier to maintain/extend in the future. For example, if we wanted to add more years later.

File: laos_gggi/GPCC_data_loader.py [hunk 2]
```diff
@@ -2,81 +2,111 @@
...
 import gzip
 import shutil
 
 
 def download_gpcc_data(output_path="../data"):
-    path_to_GPCC_raw = { '1981_1990' : output_path + "/gpcc/gpcc_raw_1981_1990.nc.gz",
-                       '1991_2000' : output_path + "/gpcc/gpcc_raw_1991_2000.nc.gz",
-                       '2001_2010':output_path + "/gpcc/gpcc_raw_2001_2010.nc.gz" ,
-                       '2011_2020':output_path + "/gpcc/gpcc_raw_2011_2020.nc.gz"}
+    path_to_GPCC_raw = {
```
[the remark this answers is not in the document]
Comment (CamSalda) [reply laos_gggi/GPCC_data_loader.py:13]: Understood! Done!

File: laos_gggi/GPCC_data_loader.py [hunk 3]
```diff
@@ -2,81 +2,111 @@
...
+        "1981_1990": output_path + "/gpcc/gpcc_raw_1981_1990.nc.gz",
+        "1991_2000": output_path + "/gpcc/gpcc_raw_1991_2000.nc.gz",
+        "2001_2010": output_path + "/gpcc/gpcc_raw_2001_2010.nc.gz",
+        "2011_2020": output_path + "/gpcc/gpcc_raw_2011_2020.nc.gz",
+    }
     path_to_GPCC_unzipped = output_path + "/gpcc/gpcc_raw_1981_1990.nc"
-    shapefile_unzipped_folder_world = output_path + "/shapefiles/wb_countries_admin0_10m"
+    shapefile_unzipped_folder_world = (
+        output_path + "/shapefiles/wb_countries_admin0_10m"
```
[the remark this answers is not in the document]
Comment (CamSalda) [reply laos_gggi/GPCC_data_loader.py:21]: I tried it, but it creates a mess in Windows with the slash direction. So every time I use it I have to correct it. Is it really necessary?

File: laos_gggi/const_vars.py [hunk 4]
```diff
@@ -52,7 +52,9 @@
     "population_density": "Population density (people per sq. km of land area)",
 }
 
-GPCC_URL = {"1981_1990": "https://opendata.dwd.de/climate_environment/GPCC/full_data_monthly_v2022/10/full_data_monthly_v2022_1981_1990_10.nc.gz",
-    "1991_2000": "https://opendata.dwd.de/climate_environment/GPCC/full_data_monthly_v2022/10/full_data_monthly_v2022_1991_2000_10.nc.gz" ,
-    "2001_2010": "https://opendata.dwd.de/climate_environment/GPCC/full_data_monthly_v2022/10/full_data_monthly_v2022_2001_2010_10.nc.gz" ,
-    "2011_2020": "https://opendata.dwd.de/climate_environment/GPCC/full_data_monthly_v2022/10/full_data_monthly_v2022_2011_2020_10.nc.gz"}
+GPCC_URL = {
```
[the remark this answers is not in the document]
Comment (CamSalda) [reply laos_gggi/const_vars.py:55]: Done!

File: laos_gggi/GPCC_data_loader.py [hunk 5]
```diff
@@ -2,81 +2,111 @@
...
-            with open(output_path + "/gpcc/gpcc_raw_2001_2010.nc", 'wb') as f_out:
+        with gzip.open(output_path + "/gpcc/gpcc_raw_2001_2010.nc.gz", "rb") as f_in:
+            with open(output_path + "/gpcc/gpcc_raw_2001_2010.nc", "wb") as f_out:
                 shutil.copyfileobj(f_in, f_out)
 
-        with gzip.open(output_path + "/gpcc/gpcc_raw_2011_2020.nc.gz", 'rb') as f_in:
-            with open(output_path + "/gpcc/gpcc_raw_2011_2020.nc", 'wb') as f_out:
+        with gzip.open(output_path + "/gpcc/gpcc_raw_2011_2020.nc.gz", "rb") as f_in:
+            with open(output_path + "/gpcc/gpcc_raw_2011_2020.nc", "wb") as f_out:
                 shutil.copyfileobj(f_in, f_out)
```
[the remark this answers is not in the document]
Comment (CamSalda) [reply laos_gggi/GPCC_data_loader.py:69]: Done! looks much cleaner!

File: laos_gggi/GPCC_data_loader.py (hunk 1 above)
Comment (CamSalda) [reply laos_gggi/GPCC_data_loader.py:82]: Understood!

Done!

File: laos_gggi/GPCC_data_loader.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (CamSalda) [reply laos_gggi/GPCC_data_loader.py:22]: @jessegrabowski I just pushed all the required changes

[the remark this answers is not in the document]
Comment (CamSalda) [reply laos_gggi/GPCC_data_loader.py:22]: Regarding os.path.join, I am also having problems with relative path variables as it does not recognize output_path variable where output_path="../data"

I tried to change it but the amount of problems it creates seems to be really big and this Issue has already consumed a massive amount of time  XD

**Review by jessegrabowski — APPROVED** [review]

**Pull request merged**